tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/scheduler/OCLNVIDIAGPUScheduler.java

```diff
@@ -0,0 +1,116 @@
+ *
+ */
+package uk.ac.manchester.tornado.drivers.opencl.scheduler;
+
+import uk.ac.manchester.tornado.drivers.opencl.OCLDeviceContext;
+import uk.ac.manchester.tornado.drivers.opencl.OCLTargetDevice;
+import uk.ac.manchester.tornado.runtime.tasks.meta.TaskMetaData;
+
+public class OCLNVIDIAGPUScheduler extends OCLKernelScheduler {
+
+    private static final int WARP_SIZE = 32;
+    private boolean ADJUST_IRREGULAR = false;
+
+    private final long[] maxWorkItemSizes;
+
+    public OCLNVIDIAGPUScheduler(final OCLDeviceContext context) {
+        super(context);
+        OCLTargetDevice device = context.getDevice();
+
+        maxWorkItemSizes = device.getDeviceMaxWorkItemSizes();
+    }
+
+    @Override
+    public void calculateGlobalWork(final TaskMetaData meta, long batchThreads) {
+        final long[] globalWork = meta.getGlobalWork();
+        for (int i = 0; i < meta.getDims(); i++) {
+            long value = (batchThreads <= 0) ? (long) (meta.getDomain().get(i).cardinality()) : batchThreads;
+            if (ADJUST_IRREGULAR && (value % WARP_SIZE != 0)) {
+                value = ((value / WARP_SIZE) + 1) * WARP_SIZE;
+            }
+            globalWork[i] = value;
+        }
+    }
+
+    @Override
+    public void calculateLocalWork(final TaskMetaData meta) {
+        final long[] localWork = meta.initLocalWork();
+        switch (meta.getDims()) {
+            case 3:
+                localWork[2] = calculateGroupSize(calculateEffectiveMaxWorkItemSizes(meta)[2], meta.getGlobalWork()[2], 3);
+                localWork[1] = calculateGroupSize(calculateEffectiveMaxWorkItemSizes(meta)[1], meta.getGlobalWork()[1], 3);
+                localWork[0] = calculateGroupSize(calculateEffectiveMaxWorkItemSizes(meta)[0], meta.getGlobalWork()[0], 3);
+                break;
+            case 2:
+                localWork[1] = calculateGroupSize(calculateEffectiveMaxWorkItemSizes(meta)[1], meta.getGlobalWork()[1], 2);
+                localWork[0] = calculateGroupSize(calculateEffectiveMaxWorkItemSizes(meta)[0], meta.getGlobalWork()[0], 2);
+                break;
+            case 1:
+                localWork[0] = calculateGroupSize(calculateEffectiveMaxWorkItemSizes(meta)[0], meta.getGlobalWork()[0], 1);
+                break;
+            default:
+                break;
+        }
+    }
+
+    private int calculateGroupSize(long maxBlockSize, long globalWorkSize, int dim) {
+        if (maxBlockSize == globalWorkSize) {
+            maxBlockSize /= 4;
+        }
+
+        int value = (int) Math.min(maxBlockSize, globalWorkSize);
+        if (value == 0) {
+            return 1;
+        }
+        while (globalWorkSize % value != 0) {
+            value--;
+        }
+        if (value >= 32 && dim > 1) {
+            value /= 2;
+        }
+        return value;
+    }
+
+    private long[] calculateEffectiveMaxWorkItemSizes(TaskMetaData metaData) {
+        long[] intermediates = new long[] { 1, 1, 1 };
```

Comment:
any particular reasoning for this particular name of the array? Shouldn't this reflect the max work items?

Reply:
Yes, I agree. I just did a refactoring of this method